contracts/HoprStakeSeason3.sol

```diff
@@ -0,0 +1,492 @@
+     * @param from address token holder address
+     * @param to address recipient address
+     * @param amount uint256 amount of tokens to transfer
+     * @param userData bytes hex information provided by the token holder (if any)
+     * @param operatorData bytes extra information provided by the operator (if any)
+     */
+    function tokensReceived(
+        // solhint-disable-next-line no-unused-vars
+        address operator,
+        address from,
+        address to,
+        uint256 amount,
+        // solhint-disable-next-line no-unused-vars
+        bytes calldata userData,
+        // solhint-disable-next-line no-unused-vars
+        bytes calldata operatorData
+    ) external override {
+        require(msg.sender == REWARD_TOKEN, "HoprStake: Sender must be wxHOPR token");
+        require(to == address(this), "HoprStake: Must be sending tokens to HoprStake contract");
+        require(from == owner(), "HoprStake: Only accept owner to provide rewards");
+        availableReward += amount;
+        emit RewardFueled(amount);
+    }
+
+    /**
+     * @dev Whenever a boost `tokenId` token is transferred to this contract via {IERC721-safeTransferFrom}
+     * when redeeming, this function is called. Boost factor associated with the 
+     * It must return its Solidity selector to confirm the token transfer upon success.
+     * @param operator address operator requesting the transfer
+     * @param from address token holder address
+     * @param tokenId uint256 amount of tokens to transfer
+     * @param data bytes hex information provided by the token holder (if any)
+     */
+    function onERC721Received(
+        // solhint-disable-next-line no-unused-vars
+        address operator,
+        address from,
+        uint256 tokenId,
+        // solhint-disable-next-line no-unused-vars
+        bytes calldata data
+    ) external override returns (bytes4) {
+        require(_msgSender() == address(NFT_CONTRACT), "HoprStake: Cannot SafeTransferFrom tokens other than HoprBoost.");
+        require(block.timestamp <= PROGRAM_END, "HoprStake: Program ended, cannot redeem boosts.");
+        // Account memory account = accounts[from];
+        _sync(from);
+
+        // redeem NFT
+        redeemedNft[from][redeemedNftIndex[from]] = tokenId;
+        redeemedNftIndex[from] += 1;
+
+        // update boost factor
+        uint256 typeId = NFT_CONTRACT.typeIndexOf(tokenId);
+        (uint256 factor, ) = NFT_CONTRACT.boostOf(tokenId);
+
+        uint256 boostIndex = redeemedFactorIndex[from];
+        uint256 index = 0;
+        for (index; index < boostIndex; index++) {
+            // loop through redeemed factors, replace the factor of the same type, if the current factor is larger.
+            uint256 redeemedId = redeemedFactor[from][index];
+            uint256 redeemedIndex = NFT_CONTRACT.typeIndexOf(redeemedId);
+            require(!isBlockedNft[redeemedIndex], "HoprStake: Can only redeem NFTs of allowed types.");
```

Comment:
This looks like it would break when an NFT is blocked at runtime when its already part of the `redeemedFactor`? Should it just be ignored instead?

Reply:
In short, yes, we can do that. 
Currently reverts the entire transaction when an NFT is blocked at runtime. If an NFT got redeemed and its type get later blocked, 1) the redeemed NFT will still be kept in the staking contract and can only be returned at the end of the program, 2) the factor is still registered and contribute to APY calculation. To be able to precisely update the APY (i.e. let the APY decrease from the time when it's blocked), either we sync for all the stakers when a type gets (un)blocked, which is not possible, or we let the contract record the timestamp when an NFT type gets blocked and modify the `_getCumulatedRewardsIncrement` function so that it checks against the latest list and computes rewards.

Reply:
Update the original issue after sync with Tino, clarifying the unnecessary complexity associated with blocking NFTs at runtime https://github.com/hoprnet/hopr-devrel/issues/205#issuecomment-1108450154